### ISSUE TITLE
oneDNN compile fail with latest sycl open source compiler for error get_pointer

### ISSUE THREAD

**Issue by jingwan2** [oneapi-src/oneDNN#1665]
# Summary
oneDNN compile fail with latest sycl open source compiler for error get_pointer

# Version
commit 2ede31e834a25ca14c648e8617b972148c94554c (HEAD -> rls-v3.1, origin/rls-v3.1)

# Environment
* OS version (5.15.0-72-generic #79-Ubuntu SMP Wed Apr 19 08:22:18 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux)
* Compiler version (clang version 17.0.0 (https://github.com/intel/llvm.git 2648b7c5e1a4af8e0ffded1c431b79813fb24777))
* CMake version (cmake version 3.15.5)
* git hash (commit 2ede31e834a25ca14c648e8617b972148c94554c (HEAD -> rls-v3.1, origin/rls-v3.1))

# Steps to reproduce
```
mkdir build
cd build
export CC=`which clang`
export CXX=`which clang++`
cmake -DCUBLAS_INCLUDE_DIR=/usr/local/cuda/include -DCUDNN_INCLUDE_DIR=/rdrive/tests/data/mainline/CT-SyclTests/benchmarks/oneBench/common/cuDNN/latest/include/ -DCUDNN_LIBRARY=/rdrive/tests/data/mainline/CT-SyclTests/benchmarks/oneBench/common/cuDNN/latest/lib/libcudnn.so -DCUBLAS_LIBRARY=/usr/local/cuda/lib64/libcublas.so -DDNNL_CPU_RUNTIME=NONE -DDNNL_GPU_RUNTIME=DPCPP -DDNNL_GPU_VENDOR=NVIDIA -DONEDNN_BUILD_GRAPH=OFF -DDNNL_BUILD_TESTS=OFF -G Ninja ..
cmake --build . -j 112
```

# Log
```
../src/gpu/sycl/sycl_types.hpp:67:72: error: member reference base type 'std::add_pointer_t<value_type>' (aka 'unsigned char *') is not a structure or union
    void *get_pointer() const { return usm_ ? usm_ : acc_.get_pointer().get(); }
                                                     ~~~~~~~~~~~~~~~~~~^~~~
../src/gpu/sycl/binary_kernels.hpp:83:41: note: in instantiation of member function 'dnnl::impl::gpu::sycl::sycl_memory_arg_t<sycl::access::mode::write>::get_pointer' requested here
    void *dst_ptr() const { return dst_.get_pointer(); }
                                        ^
../src/gpu/sycl/binary_kernels.hpp:81:43: error: SYCL kernel cannot call an undefined function without SYCL_EXTERNAL attribute
    void *src0_ptr() const { return src0_.get_pointer(); }
                                          ^
../src/gpu/sycl/sycl_types.hpp:67:11: note: 'get_pointer' declared here
    void *get_pointer() const { return usm_ ? usm_ : acc_.get_pointer().get(); }
          ^
../src/gpu/sycl/binary_kernels.hpp:81:11: note: called by 'src0_ptr'
    void *src0_ptr() const { return src0_.get_pointer(); }
          ^
../src/gpu/sycl/binary_kernels.hpp:82:43: error: SYCL kernel cannot call an undefined function without SYCL_EXTERNAL attribute
    void *src1_ptr() const { return src1_.get_pointer(); }
                                          ^
../src/gpu/sycl/sycl_types.hpp:67:11: note: 'get_pointer' declared here
    void *get_pointer() const { return usm_ ? usm_ : acc_.get_pointer().get(); }
          ^
../src/gpu/sycl/binary_kernels.hpp:82:11: note: called by 'src1_ptr'
    void *src1_ptr() const { return src1_.get_pointer(); }
          ^
../src/gpu/sycl/binary_kernels.hpp:83:41: error: SYCL kernel cannot call an undefined function without SYCL_EXTERNAL attribute
    void *dst_ptr() const { return dst_.get_pointer(); }
                                        ^
../src/gpu/sycl/sycl_types.hpp:67:11: note: 'get_pointer' declared here
    void *get_pointer() const { return usm_ ? usm_ : acc_.get_pointer().get(); }
          ^
../src/gpu/sycl/binary_kernels.hpp:83:11: note: called by 'dst_ptr'
    void *dst_ptr() const { return dst_.get_pointer(); }
          ^
../src/gpu/sycl/binary_kernels.hpp:85:49: error: SYCL kernel cannot call an undefined function without SYCL_EXTERNAL attribute
        return static_cast<float *>(src0_scale_.get_pointer());
                                                ^
../src/gpu/sycl/sycl_types.hpp:67:11: note: 'get_pointer' declared here
    void *get_pointer() const { return usm_ ? usm_ : acc_.get_pointer().get(); }
          ^
../src/gpu/sycl/binary_kernels.hpp:84:12: note: called by 'src0_scale_ptr'
    float *src0_scale_ptr() const {
           ^
../src/gpu/sycl/binary_kernels.hpp:88:49: error: SYCL kernel cannot call an undefined function without SYCL_EXTERNAL attribute
        return static_cast<float *>(src1_scale_.get_pointer());
                                                ^
../src/gpu/sycl/sycl_types.hpp:67:11: note: 'get_pointer' declared here
    void *get_pointer() const { return usm_ ? usm_ : acc_.get_pointer().get(); }
          ^
../src/gpu/sycl/binary_kernels.hpp:87:12: note: called by 'src1_scale_ptr'
    float *src1_scale_ptr() const {
           ^
7 errors generated.
```

**Comment by vpirogov** [comment]
I can reproduce the issue with https://github.com/intel/llvm/commit/485221047281e3d47f7376394667b85a63173991. This looks like a compiler bug though.

You can use either of the following options to build oneDNN for NVIDIA GPUs:
* [Intel oneAPI DPC++/C++ Compiler](https://www.intel.com/content/www/us/en/developer/tools/oneapi/dpc-compiler.html) with [NVIDIA plugin](https://developer.codeplay.com/products/oneapi/nvidia/home)
* [intel/llvm nightly 20230122](https://github.com/intel/llvm/releases/tag/sycl-nightly%2F20230122)

**Comment by jingwan2** [comment]
@vpirogov, the latest compiler implements get_pointer as SYCL 2020 spec (https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html), it's better oneDNN also follow that.

**Comment by vpirogov** [comment]
@densamoilov, does this make sense to you?

**Comment by vpirogov** [comment]
Thanks for pointer, @jingwan2. So observed behavior is the result of API breaking change in the compiler. In oneDNN we'll need to figure out how to support old behavior, which is currently in production, along with the new one.

**Comment by vpirogov** [comment]
Relevant changeset in the compiler: https://github.com/intel/llvm/pull/8874

**Comment by vpirogov** [comment]
Fixed by e5ba8f607cc5269d92019fb4b1fbfea9cf02d830.
